### ISSUE TITLE
[Usage]: vLLM For maximally batched use case

### ISSUE THREAD

**Issue by BoazTouitou97** [vllm-project/vllm#9760]
### Your current environment

```text
The output of `python collect_env.py`
```

### How would you like to use vllm

Hey, I am trying to run LLAMA 3.1 70b on an 8xL40s (PCI-E) setup. 
My use case is a single massive batch, and I want to run the run_batch endpoint with maximally large throughput. 
Since it's an offline use case I do not care about the latency much. 
Any tips as to the best parameters? Specifically what to set max_num_batched_tokens to
And is there any benefit to running it in tensor/pipeline parallel performance wise?
Would also love any thoughts on reordering the prompts (There's no shared prefix)
I've looked at #2492 and #8513

### Before submitting a new issue...

- [X] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

**Comment by ehartford** [comment]
I also need guidance for how to set all these parameters to maximize throughput on a very large offline batch job - any ideas?

**Comment by ehartford** [comment]
No...
I just ended up asking Chatgpt to generate something for me

**Comment by stefanpie** [comment]
Bumping this use case as well. Helpful for large scale LLM benchmarks and inference scaling experiments. So far I have just messed with vLLM parameters to get something that kinda is good enough.

**Comment by fxmarty-amd** [comment]
@stefanpie is your use case offline, caring only about global throughput?

I guess vllm v1 would help (https://blog.vllm.ai/2025/01/27/v1-alpha-release.html) but I had no luck running it on ROCm.

Otherwise, if your use case is indeed offline, give a try to `--block-size 32` see if it helps. Not sure whether the influence is the same in nvidia devices, but at least in my case going from block size = 16 (default on cuda & rocm) to 32 helped a lot with global throughput. Actually, even 64 allowing vllm to compile it at https://github.com/vllm-project/vllm/blob/c8d70e2437feecdb3762ce17298df33439ae1bd1/csrc/attention/paged_attention_v1.cu#L151-L167 & https://github.com/vllm-project/vllm/blob/c8d70e2437feecdb3762ce17298df33439ae1bd1/csrc/attention/paged_attention_v2.cu#L158-L174

helped a lot. I guess this is dependent on the sequence lengths though?

Doing the eval of `meta-llama/Llama-2-7b-chat-hf` through lm-eval-harness using 

```bash
lm_eval --model local-completions \
    --model_args model=meta-llama/Llama-2-7b-chat-hf,base_url=http://0.0.0.0:8000/v1/completions,num_concurrent=9999999,timeout=99999,tokenized_requests=False \
    --tasks gsm8k \
    --gen_kwargs "do_sample=False,top_p=1,temperature=0"
````

and vllm server started with

```bash
VLLM_USE_TRITON_FLASH_ATTN=0 CUDA_VISIBLE_DEVICES=6 vllm serve meta-llama/Llama-2-7b-chat-hf \
    --tensor-parallel-size 1 --num-scheduler-steps 8 --multi-step-stream-outputs=False \
    --block-size 64
```

gave me for various block sizes

| block_size | eval time | flexible-extract | strict-match |
|------------|-----------|:----------------:|:------------:|
| 16         | 19:27     |      0.2343      |    0.2328    |
| 32         | 14:26     |      0.2328      |    0.2312    |
| 64         | 05:28     |      0.2305      |     0.229    |
| 128        | 07:57     |      0.2312      |    0.2297    |

Using `torch==2.7.0.dev20250218+rocm6.3`, and rocm/vllm at 17b26bdebee9126d5486064f111d70fa8274311f.

**Comment by fxmarty-amd** [comment]
^ woops, the above influence of `--block-size` seems to be valid only for MI250, no luck on MI300 or H100.

**Comment by stefanpie** [comment]
@fxmarty-amd yeah, my use case is offline batch inference, i.e., running large LLM benchmarks.

I actually have several different benchmark cases using the LLM class to make the abstraction easier (for now):

- Different prompt -> response pairs, each with a small n_samples, ~10 per prompt (I can also "batch" multiple prompts myself when passing into LLM.chat() for inference since it's a zero-shot eval)

- Different prompt -> response pairs, each with a huge n_samples, ~100-1000 samples per prompt (I can "batch" prompts myself to pass into LLM.chat() as before; however, since n_samples here is huge, I usually just call LLM.chat() for a single prompt at a time and set n_samples to be large. I also call LLM.chat() first with a single prompt and SamplingParams(n=1, max_tokens=1) to fill the prefix cache)

- Prompt -> response pairs for tool feedback chains; since the tool runtimes are not under my control, I end up not being able to use the LLM class since I need to be able to submit prompts asynchronously and let vLLM properly batch and schedule everything on its own. I also run multiple tool feedback chains in parallel that submit prompts, so I'm not sure if you can call LLM.chat() from different threads or cores if using multiprocessing in a way that won't block other threads/cores. For this case, I am working on just using AsyncLLMEngine directly. Also note in this use case, I also need to collect a small n_samples, ~10 per prompt, for each step in the tool feedback.

I currently have gotten the first two benchmarking situations to work successfully for my research and am working on integrating AsyncLLMEngine to make the third use case work for my research as well.

I would also say my sequence lengths are ~16K at most for zero-shot cases but can maybe get longer to ~32K for tool feedback cases (these are like upper bound estimates and also depend on if the model I'm using supports those lengths).

I have access to 4 Nvidia RTX A6000 GPUs which I can use for my research, allowing me to, in most cases, set tensor_parallel_size to either 2 or 4 depending on the model.

Let me see if I can get vLLM v1 working for me since I just saw that as well. I will also try block size 32 and see if that helps.

I hope others share feedback or tips for these use cases and offline inference as well.
